### PR TITLE
Add Penn Treebank

### PR DESCRIPTION
**Pull request by youchan**
`Penn Tree Bank <https://www.cis.upenn.edu/~treebank/>`_ is originally a
corpus of English sentences with linguistic structure annotations. This
function uses a variant distributed at
`https://github.com/wojzaremba/lstm <https://github.com/wojzaremba/lstm>`_,
which omits the annotation and splits the dataset into three parts:
training, validation, and test.

c.f. https://github.com/chainer/chainer/blob/master/chainer/datasets/ptb.py

### CODE REVIEW

File: lib/datasets/penn-treebank.rb [hunk 1]
```diff
@@ -53,19 +55,17 @@ def vocabulary
...
-            x = if vocab.has_key?(word)
-              vocab[word]
-            else
+            unless vocab.key?(word)
               vocab[word] = index
-              index.tap{ index += 1 }
+              index += 1
             end
-            block.call(x)
+            yield(Record.new(word, index))
```
Comment (youchan) [inline lib/datasets/penn-treebank.rb]: `index` will be different from `vocab[word]`.

**Comment by youchan** [comment]
`vocabulary` is necessary for word2vec.

File: lib/datasets/penn-treebank.rb (hunk 1 above)
Comment (kou) [reply lib/datasets/penn-treebank.rb]: Oh, sorry.

**Comment by youchan** [comment]
I rethink it is good to store `vocabulary` by the client side.

**Comment by kou** [comment]
Do we need `{word1 => id1, word2 => id2, ...}` data structure?
We should consider generic API that can be used in many datasets for the data structre.

**Comment by kou** [comment]
For example, we already have `to_table` API. Users can get data as `{:word => [word1, word2, ...], :id => [id1, id2, ...]}` (columner) structure.
We can consider for `{word1 => id1, word2 => id2, ...}` data structure if we need the data structure.

**Comment by kou** [comment]
@youchan Do you have any other concerns for my changes?

**Comment by kou** [comment]
I have one concern.
What markup should we use for dataset description. It should be handled as a new issue.

**Comment by youchan** [comment]
We need 2 data structures.

1. full words as id list `[id1, id2, ...]`
2. vocabrary (word list not duplicated and convertible from id)

**Comment by youchan** [comment]
description from chainer/datasets

```
def get_ptb_words():
    """Gets the Penn Tree Bank dataset as long word sequences.

    `Penn Tree Bank <https://www.cis.upenn.edu/~treebank/>`_ is originally a
    corpus of English sentences with linguistic structure annotations. This
    function uses a variant distributed at
    `https://github.com/wojzaremba/lstm <https://github.com/wojzaremba/lstm>`_,
    which omits the annotation and splits the dataset into three parts:
    training, validation, and test.

    This function returns the training, validation, and test sets, each of
    which is represented as a long array of word IDs. All sentences in the
    dataset are concatenated by End-of-Sentence mark '<eos>', which is treated
    as one of the vocabulary.

    Returns:
        tuple of numpy.ndarray: Int32 vectors of word IDs.

    .. Seealso::
       Use :func:`get_ptb_words_vocabulary` to get the mapping between the
       words and word IDs.

    """
```

```
def get_ptb_words_vocabulary():
    """Gets the Penn Tree Bank word vocabulary.

    Returns:
        dict: Dictionary that maps words to corresponding word IDs. The IDs are
        used in the Penn Tree Bank long sequence datasets.

    .. seealso::
       See :func:`get_ptb_words` for the actual datasets.

    """
```

**Comment by kou** [comment]
Thanks.

> full words as id list `[id1, id2, ...]`

We can get it by `ptb.to_table[:id]`.

> vocabrary (word list not duplicated and convertible from id)

We don't have API for it. We should handle it as a new issue.

I'll merge this pull request for now. Thanks for your work!

**Comment by youchan** [comment]
`word IDs` will be treated as matrix (Numo/NArray). So we need conversion table to word from ID.
And we need the number of vocabulary.

**Comment by youchan** [comment]
Thank you!
I concern that vocabulary will be good to treat by the client.

**Comment by kou** [comment]
> description from chainer/datasets

I see why you use reStructuredText. :-)

https://www.cis.upenn.edu/~treebank/ is a dead link. Can you report it to Chainer developers?

**Comment by kou** [comment]
I've created #22. If we need to handle the use case in red-datasets, we can just close #22.

**Pull request merged**